sonar-xml-plugin/src/main/java/org/sonar/plugins/xml/LineCounter.java

```diff
@@ -66,18 +51,72 @@ private static void saveMeasures(XmlFile xmlFile, LineCountData data, FileLinesC
       .save();
   }
 
-  public static void analyse(SensorContext context, FileLinesContextFactory fileLinesContextFactory, XmlFile xmlFile) {
-    LOG.debug("Count lines in {}", xmlFile.uri());
-
-    try {
-      saveMeasures(
-        xmlFile,
-        new LineCountParser(xmlFile.getContents(), xmlFile.getCharset()).getLineCountData(),
-        fileLinesContextFactory.createFor(xmlFile.getInputFile()), context);
-    } catch (IOException | SAXException e) {
-      LOG.debug("Unable to count lines for file " + xmlFile.uri());
-      throw new ParseException(e);
+  public static void analyse(SensorContext context, FileLinesContextFactory fileLinesContextFactory, NewXmlFile xmlFile) {
+    LOG.debug("Count lines in {}", xmlFile.getInputFile().uri());
+
+    Set<Integer> linesOfCode = new HashSet<>();
+    Set<Integer> commentLines = new HashSet<>();
+
+    Document document = xmlFile.getDocument();
+    visitNode(document, linesOfCode, commentLines);
+
+    xmlFile.getPrologElement().ifPresent(prologElement ->
+      addLinesRange(
+        linesOfCode,
+        prologElement.getPrologStartLocation().getStartLine(),
+        prologElement.getPrologEndLocation().getEndLine()));
+
+    FileLinesContext fileLinesContext = fileLinesContextFactory.createFor(xmlFile.getInputFile());
+    linesOfCode.forEach(lineOfCode -> fileLinesContext.setIntValue(CoreMetrics.NCLOC_DATA_KEY, lineOfCode, 1));
+    fileLinesContext.save();
+
+    saveMeasure(context, xmlFile.getInputFile(), CoreMetrics.COMMENT_LINES, commentLines.size());
+    saveMeasure(context, xmlFile.getInputFile(), CoreMetrics.NCLOC, linesOfCode.size());
+
+  }
+
+  private static void visitNode(Node node, Set<Integer> linesOfCode, Set<Integer> commentLines) {
+    XmlTextRange range = NewXmlFile.nodeLocation(node);
+
+    switch (node.getNodeType()) {
+      case Node.ELEMENT_NODE:
+        addLinesRange(linesOfCode, NewXmlFile.startLocation((Element) node));
+        addLinesRange(linesOfCode, NewXmlFile.endLocation((Element) node));
+        break;
+      case Node.COMMENT_NODE:
+        addNotEmptyLines(commentLines, node.getTextContent(), range);
+        break;
+      case Node.TEXT_NODE:
+      case Node.CDATA_SECTION_NODE:
+      case Node.DOCUMENT_TYPE_NODE:
+        addNotEmptyLines(linesOfCode, node.getTextContent(), range);
+        break;
+      default:
+        break;
     }
+
+    NewXmlFile.children(node).forEach(child -> visitNode(child, linesOfCode, commentLines));
+  }
+
+  private static void addNotEmptyLines(Set<Integer> set, String text, XmlTextRange fullTextRange) {
+    String[] lines = text.split("(\r)?\n|\r", -1);
```

Comment:
Could you extract this REGEX as a constant? So we could use it from other places and be consistent